Runigram.java

```diff
@@ -60,48 +70,99 @@ private static void print(Color c) {
-		//// Replace this comment with your code
+		
+		for(int i=0; i< image.length; i++)
+		{
+			for(int j=0; j< image[i].length; j++)
+			{
+				print(image[i][j]);
+				System.out.print("  ");
+			}
+			System.out.println();
+		}
 	}
 	
 	/**
 	 * Returns an image which is the horizontally flipped version of the given image. 
 	 */
 	public static Color[][] flippedHorizontally(Color[][] image) {
-		//// Replace the following statement with your code
-		return null;
+		int numRows = image.length;
+		int numCols = image[0].length;
+		Color[][] flippedImage = new Color[numRows][numCols];
+		for(int i=0; i< numRows; i++)
+		{
+			for(int j=0 ; j< numCols; j++)
+			{
+				flippedImage[i][numCols-1-j] = image[i][j];
+			}
+		}
+		return flippedImage;
 	}
 	
 	/**
 	 * Returns an image which is the vertically flipped version of the given image. 
 	 */
 	public static Color[][] flippedVertically(Color[][] image){
-		//// Replace the following statement with your code
-		return null;
+		int numRows = image.length;
+		int numCols = image[0].length;
+		Color[][] flippedImage = new Color[numRows][numCols];
+		for(int i=0; i< numRows; i++)
+		{
+			for(int j=0 ; j< numCols; j++)
+			{
+				flippedImage[numRows-i-1][j] = image[i][j];
+			}
+		}
+		return flippedImage;
 	}
 	
 	// Computes the luminance of the RGB values of the given pixel, using the formula 
 	// lum = 0.299 * r + 0.587 * g + 0.114 * b, and returns a Color object consisting
 	// the three values r = lum, g = lum, b = lum.
 	public static Color luminance(Color pixel) {
-		//// Replace the following statement with your code
-		return null;
+		double red = 0.299 * pixel.getRed();
+		double green = 0.587 * pixel.getGreen();
+		double blue = 0.114 * pixel.getBlue();
+		int luminance = (int)(red + green + blue);
+		Color greyPixel = new Color(luminance, luminance, luminance);	
+
+		return greyPixel;
 	}
 	
 	/**
 	 * Returns an image which is the grayscaled version of the given image.
 	 */
 	public static Color[][] grayScaled(Color[][] image) {
-		//// Replace the following statement with your code
-		return null;
+
+		Color [][] greyImage = new Color[image.length][image[0].length];
+		for(int i=0; i< image.length; i++)
+		{
+			for(int j=0; j<image[0].length; j++)
+			{
+				greyImage[i][j] = luminance(image[i][j]);
+			}
+		}
+		return greyImage;
 	}	
 	
 	/**
 	 * Returns an image which is the scaled version of the given image. 
 	 * The image is scaled (resized) to have the given width and height.
 	 */
 	public static Color[][] scaled(Color[][] image, int width, int height) {
-		//// Replace the following statement with your code
-		return null;
+
+		Color [][] scaledImage = new Color [height][width];
+		for(int i=0; i< scaledImage.length;i++)
+		{
+			double scaleI = ((image.length)/(double)(height));
+			for(int j=0; j<scaledImage[0].length ; j++)
+			{
+				double scaleJ = ((image[0].length) /(double)(width));
+				scaledImage[i][j]= image[(int)(i * scaleI)][(int)(j* scaleJ)];
+			}
+		}

```

Comment:
These calculation are not dependent on loop iterators, you can avoid multiple calculations of same value.
```suggestion
		double scaleI = ((image.length)/(double)(height));
		double scaleJ = ((image[0].length) /(double)(width));
		for(int i=0; i< scaledImage.length;i++)
		{
			for(int j=0; j<scaledImage[0].length ; j++)
			{
				scaledImage[i][j]= image[(int)(i * scaleI)][(int)(j* scaleJ)];
			}
		}

```